File: MetodosPython/Algoritmos/NaivLoopUnrollingFour.py
'''
Algoritmo tomado del artículo https://arxiv.org/abs/1106.1347 y adaptado con la ayuda de inteligencia artificial

Integrantes del grupo:
Esteban Julián Ortega Tapie
Jhon Sebastián Zambrano Ariza
David Serna Restrepo

'''
"""

    Parámetros:
        matrizA : La primera matriz de entrada.
        matrizB : La segunda matriz de entrada.
        matrizResultado : La matriz donde se almacenará el resultado de la multiplicación.
        cantidadFilasMatrices (int): La cantidad de filas en las matrices de entrada y salida.
        cantidadColumnasMatrices (int): La cantidad de columnas en las matrices de entrada y salida.
        cantidadMaximaIteracionesFilaColumna (int): La cantidad máxima de iteraciones permitida para cada fila y columna.

    Retorna:
        matriz: La matriz resultado de la multiplicación.
    """
def naiv_loop_unrolling_four(matriz_a, matriz_b, matriz_resultado, cantidad_filas_matrices, cantidad_columnas_matrices, delimitador_maxima_iteraciones_fila_columna):
    # Verifica si la cantidad de iteraciones es divisible por 4

    if delimitador_maxima_iteraciones_fila_columna % 4 == 0:
        # Los va calculando de 4 en 4
        for i in range(cantidad_filas_matrices):
            for j in range(cantidad_columnas_matrices):
                aux = 0.0
                # Itera de 4 en 4 para calcular los productos de las matrices

                for k in range(0, delimitador_maxima_iteraciones_fila_columna, 4):
                  # Realiza las multiplicaciones de manera desenrollada

                    aux += matriz_a[i][k] * matriz_b[k][j] + matriz_a[i][k + 1] * matriz_b[k + 1][j] + matriz_a[i][k + 2] * matriz_b[k + 2][j] + matriz_a[i][k + 3] * matriz_b[k + 3][j]
                matriz_resultado[i][j] = aux #Se asigna aux en la posición [i][j]

    elif delimitador_maxima_iteraciones_fila_columna % 4 == 1:
        # Los va calculando de 4 en 4 y añade al final la posición en A y B que no se pudo calcular dentro del bucle de K
        delimitador_maxima_iteraciones_fila_columna -= 1
        for i in range(cantidad_filas_matrices):
            for j in range(cantidad_columnas_matrices):
                aux = 0.0
                for k in range(0, delimitador_maxima_iteraciones_fila_columna, 4):
                    aux += matriz_a[i][k] * matriz_b[k][j] + matriz_a[i][k + 1] * matriz_b[k + 1][j] + matriz_a[i][k + 2] * matriz_b[k + 2][j] + matriz_a[i][k + 3] * matriz_b[k + 3][j]
                
                # Calcula los productos adicionales de los elementos que no se pudieron incluir en el bucle principal

                matriz_resultado[i][j] = aux + matriz_a[i][delimitador_maxima_iteraciones_fila_columna] * matriz_b[delimitador_maxima_iteraciones_fila_columna][j]

    elif delimitador_maxima_iteraciones_fila_columna % 4 == 2:
        # Los va calculando de 4 en 4 y añade al final las dos posiciones en A y B que no se pudo calcular dentro del bucle de K
        delimitador_maxima_iteraciones_fila_columna1 = delimitador_maxima_iteraciones_fila_columna - 1
        delimitador_maxima_iteraciones_fila_columna -= 2
        for i in range(cantidad_filas_matrices):
            for j in range(cantidad_columnas_matrices):
                aux = 0.0
                for k in range(0, delimitador_maxima_iteraciones_fila_columna, 4):
                    aux += matriz_a[i][k] * matriz_b[k][j] + matriz_a[i][k + 1] * matriz_b[k + 1][j] + matriz_a[i][k + 2] * matriz_b[k + 2][j] + matriz_a[i][k + 3] * matriz_b[k + 3][j]
            
                # Calcula los productos adicionales de los elementos que no se pudieron incluir en el bucle principal

                matriz_resultado[i][j] = aux + matriz_a[i][delimitador_maxima_iteraciones_fila_columna] * matriz_b[delimitador_maxima_iteraciones_fila_columna][j] + matriz_a[i][delimitador_maxima_iteraciones_fila_columna1] * matriz_b[delimitador_maxima_iteraciones_fila_columna1][j]

    else:
        # Los va calculando de 4 en 4 y añade al final las tres posiciones en A y B que no se pudo calcular dentro del bucle de K
        delimitador_maxima_iteraciones_fila_columna -= 3
        delimitador_maxima_iteraciones_fila_columna1 = delimitador_maxima_iteraciones_fila_columna + 1
        delimitador_maxima_iteraciones_fila_columna2 = delimitador_maxima_iteraciones_fila_columna + 2
        for i in range(cantidad_filas_matrices):
            for j in range(cantidad_columnas_matrices):
                aux = 0.0
                for k in range(0, delimitador_maxima_iteraciones_fila_columna, 4):
                    aux += matriz_a[i][k] * matriz_b[k][j] + matriz_a[i][k + 1] * matriz_b[k + 1][j] + matriz_a[i][k + 2] * matriz_b[k + 2][j] + matriz_a[i][k + 3] * matriz_b[k + 3][j]
                # Calcula los productos adicionales de los elementos que no se pudieron incluir en el bucle principal

                matriz_resultado[i][j] = aux + matriz_a[i][delimitador_maxima_iteraciones_fila_columna] * matriz_b[delimitador_maxima_iteraciones_fila_columna][j] + matriz_a[i][delimitador_maxima_iteraciones_fila_columna2] * matriz_b[delimitador_maxima_iteraciones_fila_columna2][j] + matriz_a[i][delimitador_maxima_iteraciones_fila_columna1] * matriz_b[delimitador_maxima_iteraciones_fila_columna1][j]

    return matriz_resultado   #Retorna la matriz resultado

File: MetodosJavascript/Algoritmos/NaivLoopUnrollingFour.js
/*
Algoritmo tomado del artículo https://arxiv.org/abs/1106.1347 y adaptado con la ayuda de inteligencia artificial

Integrantes del grupo:
Esteban Julián Ortega Tapie
Jhon Sebastián Zambrano Ariza
David Serna Restrepo
*/

/**
 * Realiza la multiplicación de matrices utilizando un bucle naive con desenrollado de bucle de 4 iteraciones.
 *
 * @param {Array<Array<number>>} matrizA - La primera matriz a multiplicar.
 * @param {Array<Array<number>>} matrizB - La segunda matriz a multiplicar.
 * @param {Array<Array<number>>} matrizResultado - La matriz donde se almacenará el resultado de la multiplicación.
 * @param {number} cantidadFilasMatrices - La cantidad de filas en las matrices A y Resultado.
 * @param {number} cantidadColumnasMatrices - La cantidad de columnas en las matrices B y Resultado.
 * @param {number} delimitadorMaximaIteracionesFilaColumna - El delimitador para la cantidad máxima de iteraciones en el bucle de fila y columna.
 * @returns {Array<Array<number>>} - La matriz resultado después de la multiplicación.
 */
export function naiv_loop_unrolling_four(matrizA, matrizB, matrizResultado, cantidadFilasMatrices, cantidadColumnasMatrices, delimitadorMaximaIteracionesFilaColumna) {

    let aux;

    // Comprueba el valor del delimitadorMaximaIteracionesFilaColumna para determinar el desenrollado del bucle
    if (delimitadorMaximaIteracionesFilaColumna % 4 == 0) {
        // Realiza la multiplicación de matrices desenrollando el bucle de 4 en 4 iteraciones
        for (let i = 0; i < cantidadFilasMatrices; i++) {
            for (let j = 0; j < cantidadColumnasMatrices; j++) {
                aux = 0.0;
                for (let k = 0; k < delimitadorMaximaIteracionesFilaColumna; k += 4) {
                    aux += matrizA[i][k] * matrizB[k][j] + matrizA[i][k + 1] * matrizB[k + 1][j] + matrizA[i][k + 2] * matrizB[k + 2][j] + matrizA[i][k + 3] * matrizB[k + 3][j];
                }
                matrizResultado[i][j] = aux;
            }
        }
    } else if (delimitadorMaximaIteracionesFilaColumna % 4 == 1) {
        // Realiza la multiplicación de matrices desenrollando el bucle de 4 en 4 iteraciones y calcula la última posición adicional
        delimitadorMaximaIteracionesFilaColumna = delimitadorMaximaIteracionesFilaColumna - 1;
        for (let i = 0; i < cantidadFilasMatrices; i++) {
            for (let j = 0; j < cantidadColumnasMatrices; j++) {
                aux = 0.0;
                for (let k = 0; k < delimitadorMaximaIteracionesFilaColumna; k += 4) {
                    aux += matrizA[i][k] * matrizB[k][j] + matrizA[i][k + 1] * matrizB[k + 1][j] + matrizA[i][k + 2] * matrizB[k + 2][j] + matrizA[i][k + 3] * matrizB[k + 3][j];
                }
                matrizResultado[i][j] = aux + matrizA[i][delimitadorMaximaIteracionesFilaColumna] * matrizB[delimitadorMaximaIteracionesFilaColumna][j];
            }
        }
    } else if (delimitadorMaximaIteracionesFilaColumna % 4 == 2) {
        // Realiza la multiplicación de matrices desenrollando el bucle de 4 en 4 iteraciones y calcula las últimas 2 posiciones adicionales
        let delimitadorMaximaIteracionesFilaColumna1 = delimitadorMaximaIteracionesFilaColumna - 1;
        delimitadorMaximaIteracionesFilaColumna = delimitadorMaximaIteracionesFilaColumna - 2;
        for (let i = 0; i < cantidadFilasMatrices; i++) {
            for (let j = 0; j < cantidadColumnasMatrices; j++) {
                aux = 0.0;
                for (let k = 0; k < delimitadorMaximaIteracionesFilaColumna; k += 4) {
                    aux += matrizA[i][k] * matrizB[k][j] + matrizA[i][k + 1] * matrizB[k + 1][j] + matrizA[i][k + 2] * matrizB[k + 2][j] + matrizA[i][k + 3] * matrizB[k + 3][j];
                }
                matrizResultado[i][j] = aux + matrizA[i][delimitadorMaximaIteracionesFilaColumna] * matrizB[delimitadorMaximaIteracionesFilaColumna][j] + matrizA[i][delimitadorMaximaIteracionesFilaColumna1] * matrizB[delimitadorMaximaIteracionesFilaColumna1][j];
            }
        }
    } else {
        // Realiza la multiplicación de matrices desenrollando el bucle de 4 en 4 iteraciones y calcula las últimas 3 posiciones adicionales
        delimitadorMaximaIteracionesFilaColumna = delimitadorMaximaIteracionesFilaColumna - 3;
        let delimitadorMaximaIteracionesFilaColumna1 = delimitadorMaximaIteracionesFilaColumna - 1;
        let delimitadorMaximaIteracionesFilaColumna2 = delimitadorMaximaIteracionesFilaColumna - 2;
        for (let i = 0; i < cantidadFilasMatrices; i++) {
            for (let j = 0; j < cantidadColumnasMatrices; j++) {
                aux = 0.0;
                for (let k = 0; k < delimitadorMaximaIteracionesFilaColumna; k += 4) {
                    aux += matrizA[i][k] * matrizB[k][j] + matrizA[i][k + 1] * matrizB[k + 1][j] + matrizA[i][k + 2] * matrizB[k + 2][j] + matrizA[i][k + 3] * matrizB[k + 3][j];
                }
                matrizResultado[i][j] = aux + matrizA[i][delimitadorMaximaIteracionesFilaColumna] * matrizB[delimitadorMaximaIteracionesFilaColumna][j] + matrizA[i][
                    delimitadorMaximaIteracionesFilaColumna2] * matrizB[delimitadorMaximaIteracionesFilaColumna2][j] + matrizA[i][delimitadorMaximaIteracionesFilaColumna1] * matrizB[delimitadorMaximaIteracionesFilaColumna1][j];
                }
            }
        }
    
        return matrizResultado;
    }
    